src/main/java/bio/terra/service/snapshot/SnapshotDao.java

```diff
@@ -48,23 +51,86 @@ public SnapshotDao(NamedParameterJdbcTemplate jdbcTemplate,
         this.datasetDao = datasetDao;
     }
 
-    @Transactional(propagation = Propagation.REQUIRED)
-    public UUID create(Snapshot snapshot) {
-        logger.debug("create snapshot " + snapshot.getName());
-        String sql = "INSERT INTO snapshot (name, description, profile_id)" +
-                " VALUES (:name, :description, :profile_id)";
+    /**
+     * Lock the snapshot object before doing something with it (e.g. delete).
+     * This method returns successfully when there is a snapshot object locked by this flight, and throws an exception
+     * in all other cases. So, multiple locks can succeed with no errors. Logic flow of the method:
+     *     1. Update the snapshot record to give this flight the lock.
+     *     2. Throw an exception if no records were updated.
+     * @param snapshotName name of the snapshot to lock, this is a unique column
+     * @param flightId flight id that wants to lock the snapshot
+     * @throws SnapshotLockException if the snapshot is locked by another flight or does not exist
+     */
+    @Transactional(propagation =  Propagation.REQUIRED, isolation = Isolation.SERIALIZABLE)
+    public void lock(String snapshotName, String flightId) {
+        if (flightId == null) {
+            throw new SnapshotLockException("Locking flight id cannot be null");
```

Comment:
SnapshotLockException is a 400. I think this error should be a 500; it seems like a coding error on our part, not an expected occurrence. Yes?

Reply:
Agreed. Changed both the Dataset and SnapshotLockException classes to extend InternalServerError instead of BadRequest.